sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala

```diff
@@ -892,11 +892,40 @@ case class MergeIntoTable(
     case _ => false
   }
 
+  private lazy val sourceSchemaForEvolution: StructType =
+    MergeIntoTable.sourceSchemaForSchemaEvolution(this)
+
+  lazy val schemaChangesNonEmpty: Boolean =
+    MergeIntoTable.schemaChanges(targetTable.schema, sourceSchemaForEvolution).nonEmpty
+
   lazy val needSchemaEvolution: Boolean =
     schemaEvolutionEnabled &&
-      MergeIntoTable.schemaChanges(targetTable.schema, sourceTable.schema).nonEmpty
+      canEvaluateSchemaEvolution &&
+      schemaChangesNonEmpty
```

Comment:
do we really need this condition? If there is no schema change, then the rule `ResolveMergeIntoSchemaEvolution` returns the original merge command. If we need to do schema change, we calculate it twice with the current code.

Reply:
moved.  Actually future rules still need this condition to run only if no schema evolution needed, so keep the flag for those.  But simplified the ResovleMergeIntoSchemaEvolution condition